cloudpickle/cloudpickle.py

```diff
@@ -79,7 +79,20 @@
 
 
 # caches dynamic modules that are not referenced in sys.modules
-_dynamic_modules_globals = {}
+_dynamic_modules_globals = weakref.WeakValueDictionary()
+
+# _dynamic_modules_globals will store _base_globals variables.
+# Or, _base_globals end up being dicts, and built-in dict instances
+# cannot be weakly referenced. Therefore, we create a mirror of the
+# dict type, and at the mutation of _base globals from string to dict,
+# we use the DynamicDict constructor instead of the builtin dict one.
+
+
+class DynamicDict(dict):
+    """class used to instanciate base_globals, in order to be weakly refrenced
+    into _dynamic_modules_globals
+    """
+    pass
```

Comment:
Why don't you keep a reference to the module object itself as I suggested in https://github.com/cloudpipe/cloudpickle/pull/201#issuecomment-420991775 ?

That would make the code much more homogeneous, no?

Reply:
In this particular case, a module was never properly created in the previous code, simply a `dict` object was created. But we can change this by using `imp.new_module` instead.

Reply:
No my point would be to register the dynamic modules themselves in `_dynamic_modules` when we unpickle them (in `dynamic_subimport` I think) and then use their name as value for `base_globals` exactly as we do for regular modules. Falling back to empty dict might still be useful for weird modules that are neither found in `sys.modules` nor registered in `_dynamic_modules`.

Reply:
Here, it is not possible, because only `mod.f`  (a function) is unpickled, so a dynamic module is never unpickled.
Creating a new module as I said does not work though, because it will only be referenced in _dynamic_modules_globals, however, since it is a `WeakValueDictionnary` , the module will be garbage collected as soon as we get out of the function where it is created.

In the current PR code, things work because ` f.__globals__ ` is a reference to a value in `_dynamic_modules_globals`, the item in this dict was not garbage collected as long as the function was not deleted, which is not the case anymore if we use a module, as f does not have a reference to the module itself.

So in this particular use case, it seems hard to keep a reference to the dynamic module where an unpickled function lived.

Reply:
Alright, that's a good point.